metadata/src/from_into/v14.rs

```diff
@@ -178,17 +226,375 @@ fn v14_to_v15(metadata: v14::RuntimeMetadataV14) -> v15::RuntimeMetadataV15 {
+
+        let Some(address) = params.get(ADDRESS) else {
+            return Err("Missing address type from extrinsic".into());
+        };
+        let Some(call) = params.get(CALL) else {
+            return Err("Missing call type from extrinsic".into());
+        };
+        let Some(signature) = params.get(SIGNATURE) else {
+            return Err("Missing signature type from extrinsic".into());
+        };
+        let Some(extra) = params.get(EXTRA) else {
+            return Err("Missing extra type from extrinsic".into());
+        };
+
+        Ok(ExtrinsicPartTypeIds {
+            address: *address,
+            call: *call,
+            signature: *signature,
+            extra: *extra,
+        })
+    }
+}
+
+fn generate_outer_enums(
+    metadata: &mut v14::RuntimeMetadataV14,
+) -> v15::OuterEnums<scale_info::form::PortableForm> {
+    let call_enum = metadata
+        .types
+        .types
+        .iter()
+        .find(|ty| {
+            let Some(ident) = ty.ty.path.ident() else { return false };
+            ident == "RuntimeCall"
+        })
+        .expect("RuntimeCall exists in V14; qed");
+
+    let event_enum = metadata
+        .types
+        .types
+        .iter()
+        .find(|ty| {
+            let Some(ident) = ty.ty.path.ident() else { return false };
+            ident == "RuntimeEvent"
+        })
+        .expect("RuntimeEvent exists in V14; qed");
+
+    let call_ty = call_enum.id.into();
+    let event_ty = event_enum.id.into();
+
+    let mut path_segments = call_enum.ty.path.segments.clone();
+    let last = path_segments
+        .last_mut()
+        .expect("Should have at least one segment checked above; qed");
+    *last = "RuntimeError".to_string();
+
+    let error_ty_id = generate_runtime_error_type(metadata, path_segments);
+
+    v15::OuterEnums {
+        call_enum_ty: call_ty,
+        event_enum_ty: event_ty,
+        error_enum_ty: error_ty_id.into(),
+    }
+}
+
+/// Generate the `RuntimeError` type and add it to the metadata.
+///
+/// Returns the `RuntimeError` Id from the registry.
+fn generate_runtime_error_type(
+    metadata: &mut v14::RuntimeMetadataV14,
+    path_segments: Vec<String>,
+) -> u32 {
+    let variants: Vec<_> = metadata
+        .pallets
+        .iter()
+        .filter_map(|pallet| {
+            let Some(pallet_error) = &pallet.error else { return None };
+
+            let error_ty = metadata
+                .types
+                .resolve(pallet_error.ty.id)
+                .expect("Error type is present in the metadata; qed");
+
+            // Note: the type name is the name of the type as it appears in substrate, we cannot
+            // accurately reproduce it when it contains more than 1 generic.
+            // The first generic is always the `Runtime`, the others represent instances.
+            let instances = ",Instance".repeat(error_ty.type_params.len() - 1);
+            let path = format!("{}<Runtime{}>", error_ty.path, instances);
```

Comment:
This looks super hacky, and I think it's just to populate the "type_name" thing? My question is; do we actually need this? Would anything break if we set `type_name` to be eg `{PalletName}Error`?

Reply:
I believe we could use the `PalletError` name here, have removed it :D

Reply:
Ok awesome! Yeah, I had a look and `type_name` isn't really relevant; it's just for "disgnostic" info: https://github.com/paritytech/scale-info/blob/cc97f4c6496acf0f92d9caa45856331f094a7f38/src/ty/fields.rs#L55

The `name` of the field is the enum variant and the thing that decoding/encoding cares about :)